src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableArray`1.cs

```diff
@@ -556,13 +559,25 @@ public ImmutableArray<T> InsertRange(int index, IEnumerable<T> items)
             }
 
             T[] tmp = new T[self.Length + count];
-            Array.Copy(self.array, 0, tmp, 0, index);
-            int sequenceIndex = index;
-            foreach (var item in items)
+            
+            // CopyTo first in case bad interface impl
+            // tries to modify other parts of tmp
+            if (items is ICollection<T>)
             {
-                tmp[sequenceIndex++] = item;
+                ((ICollection<T>)items).CopyTo(tmp, index);
```

Comment:
You now have two type checks here: the one in the if condition and the type cast here. While there is a *small* perf improvement for only doing the type check once, I doubt it's going to be significant for this method. But I'm more interested in the readability/maintainability of the code if you use `as` for the cast above the `if` line, and then the `if` check can be whether the result is non-null. That way one line of code being changed can't result in accidental `InvalidCastException`s down the road.